src/server/api/index.js

```diff
@@ -606,6 +642,48 @@ function createAPI(app, mountPath, middlewareOpts) {
             });
     });
 
+    router.get('/projects/:ownerId/:projectName/commits/:commitHash', ensureAuthenticated, function (req, res, next) {
+        var userId = getUserId(req),
+            commitHash = StorageUtil.getHashTaggedHash(req.params.commitHash),
+            data = {
+                username: userId,
+                projectId: StorageUtil.getProjectIdFromOwnerIdAndProjectName(req.params.ownerId,
+                    req.params.projectName),
+                before: commitHash,
+                number: 1
+            };
+
+        safeStorage.getCommits(data)
+            .then(function (result) {
+                res.json(result[0]);
+            })
+            .catch(function (err) {
+                if (err.message.indexOf('not exist') > -1 || err.message.indexOf('Not authorized to read') > -1 ) {
+                    err.status = 404;
+                }
+                next(err);
+            });
+    });
+
+    router.get('/projects/:ownerId/:projectName/commits/:commitHash/tree/*', ensureAuthenticated,
+        function (req, res, next) {
+            var userId = getUserId(req),
+                projectId = StorageUtil.getProjectIdFromOwnerIdAndProjectName(req.params.ownerId,
+                    req.params.projectName),
+                commitHash = StorageUtil.getHashTaggedHash(req.params.commitHash);
+
+            loadNodePathByCommitHash(userId, projectId, commitHash, '/' + req.params[0])
+                .then(function (nodeObj) {
+                    res.json(nodeObj);
+                })
+                .catch(function (err) {
+                    if (err.message.indexOf('not exist') > -1 || err.message.indexOf('Not authorized to read') > -1 ) {
+                        err.status = 404;
```

Comment:
It was requested that if the user doesn't have read access the return code should be 404.
For two reasons;
The storage just checks if the user has the project in his/her project list (with read access) and does't check whether the project exists or not.
If the user doesn't have read access, he shouldn't even be able to get information about whether the project exists or not.